pytest_rts/utils/mappinghelper.py

```diff
@@ -0,0 +1,302 @@
+        self.connection.execute(
+            """CREATE TABLE test_map (
+                    file_id INTEGER,
+                    test_function_id INTEGER,
+                    line_id INTEGER,
+                    UNIQUE(file_id,test_function_id,line_id))"""
+        )
+        self.connection.execute(
+            "CREATE TABLE src_file (id INTEGER PRIMARY KEY, path TEXT, UNIQUE (path))"
+        )
+        self.connection.execute(
+            "CREATE TABLE test_file (id INTEGER PRIMARY KEY, path TEXT, UNIQUE (path))"
+        )
+        self.connection.execute(
+            """CREATE TABLE test_function (
+                                    id INTEGER PRIMARY KEY,
+                                    test_file_id INTEGER,
+                                    context TEXT,
+                                    start INTEGER,
+                                    end INTEGER,
+                                    duration REAL,
+                                    FOREIGN KEY (test_file_id) REFERENCES test_file(id),
+                                    UNIQUE (context))"""
+        )
+        self.connection.execute("CREATE TABLE new_tests (context TEXT)")
+        self.connection.execute("CREATE TABLE last_update_hash (hash TEXT)")
+
+    def line_exists(self, file_id, line_number):
+        """Check whether a specific line number is covered for a source code file"""
+        return bool(
+            self.connection.execute(
+                "SELECT EXISTS(SELECT file_id FROM test_map WHERE file_id = ? AND line_id = ?)",
+                (
+                    file_id,
+                    line_number,
+                ),
+            ).fetchone()[0]
+        )
+
+    def __update_test_function_mapping(self, line_map, file_id):
+        """Update test function entries in the mapping database
+        by shifting their start and end line numbers
+        """
+        tests_to_update = []
+        db_data = self.connection.execute(
+            """SELECT id,test_file_id,context,start,end FROM test_function
+                WHERE test_file_id = ?""",
+            (file_id,),
+        )
+        for line in db_data:
+            tests_to_update.append(line)
+        self.connection.execute(
+            "DELETE FROM test_function WHERE test_file_id = ?", (file_id,)
+        )
+
+        for test in tests_to_update:
+            start = test[3]
+            end = test[4]
+            if start in line_map:
+                start = line_map[start]
+            if end in line_map:
+                end = line_map[end]
+            updated_test = (test[0], test[1], test[2], start, end)
+            self.connection.execute(
+                """
+                INSERT OR IGNORE
+                INTO test_function (id,test_file_id,context,start,end)
+                VALUES (?,?,?,?,?)""",
+                updated_test,
+            )
+
+    def __update_line_mapping(self, line_map, file_id):
```

Comment:
What's the reason to use double underscore?

Reply:
(here and elsewhere)

Reply:
I thought it was like a way to define a private method not meant to be called from outside of the class

Reply:
There are no real private methods in Python, whatever you do, everything will be accessible, with different difficulty. Python *convention* is to start "private" methods names with one underscore. The linter will enforce "privacy" for such methods.

Two underscores have special meaning and are very rarely used. Don't do this unless you have a really good reason.

Ref: https://docs.python.org/3/tutorial/classes.html#private-variables